weather_mv/loader_pipeline/streaming.py

```diff
@@ -0,0 +1,123 @@
+                | "Window into fixed intervals" >> beam.WindowInto(FixedWindows(self.window_size))
+                | "Add timestamp to windowed elements" >> beam.ParDo(AddTimestamp())
+                # Assign a random key to each windowed element based on the number of shards.
+                | "Add key" >> beam.WithKeys(lambda _: random.randint(0, self.num_shards - 1))
+                # Group windowed elements by key. All the elements in the same window must fit
+                # memory for this. If not, you need to use `beam.util.BatchElements`.
+                | "Group by key" >> beam.GroupByKey()
+        )
+
+
+class AddTimestamp(beam.DoFn):
+    """Processes each windowed element by extracting the message body and its
+    publish time into a tuple.
+    """
+
+    def process(self, element, publish_time=beam.DoFn.TimestampParam) -> t.Iterable[t.Tuple[str, str]]:
+        yield (
+            element.decode("utf-8"),
+            datetime.datetime.utcfromtimestamp(float(publish_time)).strftime(
+                "%Y-%m-%d %H:%M:%S.%f"
+            ),
+        )
+
+
+class ParsePaths(beam.DoFn):
+    """Parse paths to real-time weather data from windowed-batches."""
+
+    def __init__(self, uri_pattern: str):
+        self.uri_pattern = uri_pattern
+        self.protocol = f'{urlparse(uri_pattern).scheme}://'
+        super().__init__()
+
+    @classmethod
+    def try_parse_message(cls, message_body: t.Union[str, t.Dict]) -> t.Dict:
+        """Robustly parse message body, which will be JSON in the vast majority of cases,
+         but might be a dictionary."""
+        try:
+            return json.loads(message_body)
+        except (json.JSONDecodeError, TypeError):
+            if type(message_body) is dict:
+                return message_body
+            raise
+
+    def to_object_path(self, payload: t.Dict) -> str:
+        """Parse cloud object from Pub/Sub topic payload."""
+        return f'{self.protocol}{payload["bucket"]}/{payload["name"]}'
+
+    def should_skip(self, message_body: t.Dict) -> bool:
+        """Returns true if Pub/Sub topic does *not* match the target file URI pattern."""
+        try:
+            return not fnmatch.fnmatch(self.to_object_path(message_body), self.uri_pattern)
+        except KeyError:
+            return True
+
+    def process(self, key_value, window=beam.DoFn.WindowParam) -> t.Iterable[str]:
+        """Yield paths to real-time weather data in cloud storage."""
+
+        shard_id, batch = key_value
+
+        logger.debug(f'Processing shard {shard_id!r}.')
+
+        for message_body, publish_time in batch:
+            logger.debug(message_body)
+
+            parsed_msg = self.try_parse_message(message_body)
+
+            target = self.to_object_path(parsed_msg)
+            logger.info(f'Parsed path {target!r}...')
+
+            if self.should_skip(parsed_msg):
+                logger.info('skipping.')
```

Comment:
maybe add target in the log line here as well ?

Reply:
I was on the fence about adding the target, because I thought it cluttered the log messages. This is what they look like now: 
```
2022-01-05 15:18:58.114 PST Parsed path 'gs://anthromet-external-ecmwf-realtime/tmp/A1D01010000010100001.bz2'...
Info
2022-01-05 15:18:58.114 PST skipping.
Info
2022-01-05 15:19:34.746 PST Parsed path 'gs://anthromet-external-ecmwf-realtime/tmp/A1D01010000010100001.bz2'...
Info
2022-01-05 15:19:34.746 PST skipping.
Info
2022-01-05 15:22:36.700 PST Parsed path 'gs://anthromet-external-ecmwf-realtime/A1D01010000010100001.bz2'...
Info
2022-01-05 15:22:36.700 PST Extracting rows as dicts: 'gs://anthromet-external-ecmwf-realtime/A1D01010000010100001.bz2'.
```

To me, I originally though, it's clear that skipping applies to the previous parsed path. However, I now realize: If many messages come in all at once, it would be unclear what paths are being skipped (log statements will be interleaved).

This is a long way to go about saying: I agree and appreciate this note.